### PR TITLE
report: do not show element screenshot if out of bounds

### LINKED ISSUE #11121: full-page-screenshot: handle elements beyond MAX_SCREENSHOT_HEIGHT
ref https://github.com/GoogleChrome/lighthouse/pull/10716#issuecomment-657904592 and #11118 

i made https://output.jsbin.com/koqamuh/4/quiet as a test page for us. it has a (CSS) pixel ruler running down the side.

Here's a report of the page: https://marked-cable.surge.sh/color-contrast-over16000px-koqamuh.html

The elem screenshots start failing between 6200 and 6300 px

image

---------

#### DPR stuff..

6300  is  roughly 16384 / 2.625.. so  it's to do with some DPR adjustments.. (I'm capturing this on my imacpro with a native dpr of 2.)

FWIW i tried removing the `/ dpr` in maxScreenshotHeight: 
https://github.com/GoogleChrome/lighthouse/blob/06d3e1b8d69c9b7eb0a9b94c08632b097eb1eeec/lighthouse-core/gather/gatherers/full-page-screenshot.js#L75

and after doing that, [it cut out](https://marked-cable.surge.sh/color-contrast-over16000px-koqamuh-nodprdivisor.html) at 16300 . (right before 16384, the real max)

image

that much seems good.
of course, the placement of the markers are off because of some other dpr math I didnt account for.

### PR DESCRIPTION
**Pull request by connorjclark**
ref #11121

image

### CODE REVIEW

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 1]
```diff
@@ -17,6 +17,22 @@
 /** @typedef {LH.Artifacts.Rect} Rect */
 /** @typedef {{width: number, height: number}} Size */
 
+/**
+ * Returns whether rect2 is contained entirely within rect1;
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:21]: > Returns whether rect2 is contained entirely within rect1

should it not have a screenshot even if it's partially in bounds? e.g. what if just a few pixels are over the edge?

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js [hunk 2]
```diff
@@ -188,14 +207,28 @@ class ElementScreenshotRenderer {
...
   static render(dom, templateContext, fullPageScreenshot, elementRectSC, maxRenderSizeDC) {
+    const fullPageScreenshotRect = {
+      left: 0,
+      top: 0,
+      right: fullPageScreenshot.width,
+      bottom: fullPageScreenshot.height,
+      width: fullPageScreenshot.width,
+      height: fullPageScreenshot.height,
+    };
+    if (!rectContains(fullPageScreenshotRect, elementRectSC)) {
```
Comment (brendankenny) [inline lighthouse-core/report/html/renderer/element-screenshot-renderer.js:227]: replace `rectContains(rect1, rect2)` with

```js
function screenshotContainsRect(fullPageScreenshot, rect) {
  return rect.top >= 0 &&
    rect.right <= fullPageScreenshot.width &&
    rect.bottom <= fullPageScreenshot.height &&
    rect2.left >= 0;
}
```

and avoid having to setup this `fullPageScreenshotRect` object?

File: lighthouse-core/test/report/html/renderer/element-screenshot-renderer-test.js [hunk 3]
```diff
@@ -38,17 +50,17 @@ describe('ElementScreenshotRenderer', () => {
     global.Util = undefined;
   });
 
-  it('renders screenshot', () => {
+  it.only('renders screenshot', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/report/html/renderer/element-screenshot-renderer-test.js:53]:
```suggestion
  it('renders screenshot', () => {
```

**Review by brendankenny — COMMENTED** [review]
> Fixes #11121

also you'll have to hash out if this DPR stuff is resolved with @patrickhulce and/or @paulirish if we want to close that :)

File: lighthouse-core/report/html/renderer/element-screenshot-renderer.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js]: eh, I'm inclined to keep it like this as it's simpler and this seems like an edge case (hehe).

Comment (brendankenny) [reply lighthouse-core/report/html/renderer/element-screenshot-renderer.js]: fair enough :)

**Review by brendankenny — APPROVED** [review]
This change LGTM, whether or not it fully resolves #11121. Maybe we should land this but leave that issue open until all parties agree and/or we discuss next eng sync?

**Pull request merged**